### PR TITLE
merge the additional yaml config to the main yaml config

### PR DESCRIPTION
**Pull request by Nikita-Levuskin**
### Changelog entry 

Added the ability to merge additional config to the main config

### Changelog category 

* New feature

### Additional information

When starting a container without configs, you can specify the environment variable "YDB_CONFIG_YAML_PATCH", which should point to the config file relative to the workdir(default "ydb_data") and, if this variable is present, the additional config is combined with the main config. If there is no file specified in the environment variable, the application will not start

### CODE REVIEW

File: ydb/tests/library/harness/kikimr_config.py [hunk 1]
```diff
@@ -432,7 +432,32 @@ def __init__(
 
             self.yaml_config["feature_flags"]["enable_external_data_sources"] = True
             self.yaml_config["feature_flags"]["enable_script_execution_operations"] = True
-
+            
+        if os.getenv("YAML_CONFIG") is not None:
```
Comment (rekby) [inline ydb/tests/library/harness/kikimr_config.py:436]: need prefix YDB_ - for prevent conflict with other systems, because environment variables are global for all processes.

File: ydb/tests/library/harness/kikimr_config.py [hunk 2]
```diff
@@ -432,7 +432,32 @@ def __init__(
 
             self.yaml_config["feature_flags"]["enable_external_data_sources"] = True
             self.yaml_config["feature_flags"]["enable_script_execution_operations"] = True
-
+            
+        if os.getenv("YAML_CONFIG") is not None:      
+            with open("/ydb_data/" + os.getenv("YAML_CONFIG")) as fh:
```
Comment (rekby) [inline ydb/tests/library/harness/kikimr_config.py:437]: use os.path.join instead of strings concatenation. In the case it will allow set full path instead of relative path/filename.

In other cases special function may add security, for example escape special symbols. Strings concatenation usually is last way.

File: ydb/tests/library/harness/kikimr_config.py [hunk 3]
```diff
@@ -432,7 +432,32 @@ def __init__(
...
+                self.additional_yaml_config = yaml.load(fh, Loader=yaml.FullLoader)
+
+            self.yaml_config = self.merge_two_dicts(self.yaml_config, self.additional_yaml_config)
+
+    def merge_two_dicts(self, data_1, data_2):
```
Comment (rekby) [inline ydb/tests/library/harness/kikimr_config.py:442]: Did you see other way to merge dicts? May be standard library or some languages features?

File: ydb/tests/library/harness/kikimr_config.py (hunk 2 above)
Comment (Nikita-Levuskin) [reply ydb/tests/library/harness/kikimr_config.py]: fixed

File: ydb/tests/library/harness/kikimr_config.py [hunk 4]
```diff
@@ -432,7 +432,32 @@ def __init__(
 
             self.yaml_config["feature_flags"]["enable_external_data_sources"] = True
             self.yaml_config["feature_flags"]["enable_script_execution_operations"] = True
-
+            
+        if os.getenv("YDB_YAML_CONFIG") is not None:
```
Comment (rekby) [inline ydb/tests/library/harness/kikimr_config.py:436]: It will be better to move the merge code from the generator constructor to deploy function within `cmds` dir. Because merge is not part of config generator

File: ydb/public/tools/lib/cmds/__init__.py [hunk 5]
```diff
@@ -374,6 +385,10 @@ def deploy(arguments):
         **optionals
     )
 
+    if os.getenv("YDB_YAML_CONFIG") is not None:
```
Comment (asmyasnikov) [inline ydb/public/tools/lib/cmds/__init__.py:388]: YDB_YAML_CONFIG_PATCH ?

Comment (Nikita-Levuskin) [reply ydb/public/tools/lib/cmds/__init__.py]: fixed

**Review by rekby — DISMISSED** [review]

File: ydb/public/tools/lib/cmds/__init__.py [hunk 6]
```diff
@@ -309,6 +310,34 @@ def pq_client_service_types(arguments):
 def enable_pqcd(arguments):
     return (getattr(arguments, 'enable_pqcd', False) or os.getenv('YDB_ENABLE_PQCD') == 'true')
 
+def merge_two_yaml_configs(data_1, data_2): # library mergedeep did not have a version for PY2, library deepmerge didn't have the necessary functionality
+    if isinstance(data_1, dict) and isinstance(data_2, dict):
```
Comment (rekby) [inline ydb/public/tools/lib/cmds/__init__.py:314]: What if data_1/2 will not be dict/None?

File: ydb/public/tools/lib/cmds/__init__.py [hunk 7]
```diff
@@ -309,6 +310,34 @@ def pq_client_service_types(arguments):
...
+        new_dict = {}
+        d2_keys = list(data_2.keys())
+        for d1k in data_1.keys():
+            if d1k in d2_keys:
+                d2_keys.remove(d1k)
+                new_dict[d1k] = merge_two_yaml_configs(data_1.get(d1k), data_2.get(d1k))
+            else:
+                new_dict[d1k] = data_1.get(d1k)
```
Comment (rekby) [inline ydb/public/tools/lib/cmds/__init__.py:322]: Use deep copy instead of simple assign for prevent difficult mistakes.

File: ydb/public/tools/lib/cmds/__init__.py [hunk 8]
```diff
@@ -309,6 +310,34 @@ def pq_client_service_types(arguments):
...
+                for d2k in d2_keys:
```
Comment (rekby) [inline ydb/public/tools/lib/cmds/__init__.py:323]: check the indent

File: ydb/public/tools/lib/cmds/ut/test.py [hunk 9]
```diff
@@ -24,3 +24,16 @@ def test_kikimr_config_generator_generic_connector_config():
 
     actual = generic_connector_config()
     assert actual == expected
+
+def test_merge_two_yaml_configs():
```
Comment (rekby) [inline ydb/public/tools/lib/cmds/ut/test.py:28]: it needs better test own merge function

**Review by rekby — CHANGES_REQUESTED** [review]

**Review by asmyasnikov — DISMISSED** [review]

File: ydb/public/tools/lib/cmds/__init__.py [hunk 10]
```diff
@@ -310,6 +311,50 @@ def enable_pqcd(arguments):
     return (getattr(arguments, 'enable_pqcd', False) or os.getenv('YDB_ENABLE_PQCD') == 'true')
 
 
+def merge_two_yaml_configs(data_1, data_2):
+    if not (isinstance(data_1, (dict, list)) or data_1 == None):
```
Comment (rekby) [inline ydb/public/tools/lib/cmds/__init__.py:315]: it needs to check equal to None with is.
https://peps.python.org/pep-0008/#programming-recommendations

File: ydb/public/tools/lib/cmds/__init__.py [hunk 11]
```diff
@@ -310,6 +311,50 @@ def enable_pqcd(arguments):
...
+                new_dict[d1k] = merge_two_yaml_configs(data_1.get(d1k), data_2.get(d1k))
+            else:
+                new_dict[d1k] = data_1.get(d1k)
+                        
+        for d2k in d2_keys:
+            new_dict[d2k] = data_2.get(d2k)
+
+        return new_dict
+    else:
+        if data_2 == None:
```
Comment (rekby) [inline ydb/public/tools/lib/cmds/__init__.py:335]: check with `is`

File: ydb/public/tools/lib/cmds/__init__.py [hunk 12]
```diff
@@ -310,6 +311,50 @@ def enable_pqcd(arguments):
...
+            return data_1
+        else:
+            return data_2
+
+
+def check_types_for_merge(data_1, data_2):
+    if isinstance(data_1, dict) and isinstance(data_2, list):
```
Comment (rekby) [inline ydb/public/tools/lib/cmds/__init__.py:342]: It should check allowable combinations and raise exception for all other.

File: ydb/public/tools/lib/cmds/ut/test.py [hunk 13]
```diff
@@ -24,3 +24,17 @@ def test_kikimr_config_generator_generic_connector_config():
...
```
Comment (rekby) [inline ydb/public/tools/lib/cmds/ut/test.py:28]: It needs to test differnt cases with unit tests: dicts, lists, dict in list, list in dict, combinations with None, etc.

Canionization only one of tests

File: ydb/public/tools/lib/cmds/__init__.py (hunk 10 above)
Comment (rekby) [inline ydb/public/tools/lib/cmds/__init__.py:315]: It is same check as check_types_for_merge: we can't merge a list and a dict

File: ydb/public/tools/lib/cmds/__init__.py [hunk 14]
```diff
@@ -310,6 +311,50 @@ def enable_pqcd(arguments):
...
+
+        return new_dict
+    else:
+        if data_2 == None:
+            return data_1
+        else:
+            return data_2
+
+
+def check_types_for_merge(data_1, data_2):
```
Comment (rekby) [inline ydb/public/tools/lib/cmds/__init__.py:341]: it is helper function. It will be better to rename it to private variant (with prefix _).

File: ydb/public/tools/lib/cmds/__init__.py [hunk 15]
```diff
@@ -310,6 +312,54 @@ def enable_pqcd(arguments):
...
+    if isinstance(data_1, dict) and isinstance(data_2, dict):
+        data_1, data_2 = data_1.copy(), data_2.copy()
+        new_dict = {}
+        d2_keys = list(data_2.keys())
+        for d1k in data_1.keys():
+            if d1k in d2_keys:
+                d2_keys.remove(d1k)
+                new_dict[d1k] = merge_two_yaml_configs(data_1.get(d1k), data_2.get(d1k))
+            else:
+                new_dict[d1k] = data_1.get(d1k)
```
Comment (rekby) [inline ydb/public/tools/lib/cmds/__init__.py:326]: copy

File: ydb/public/tools/lib/cmds/__init__.py [hunk 16]
```diff
@@ -310,6 +312,54 @@ def enable_pqcd(arguments):
...
+                        
+        for d2k in d2_keys:
+            new_dict[d2k] = data_2.get(d2k)
```
Comment (rekby) [inline ydb/public/tools/lib/cmds/__init__.py:329]: copy

File: ydb/public/tools/lib/cmds/ut/test.py [hunk 17]
```diff
@@ -24,3 +24,45 @@ def test_kikimr_config_generator_generic_connector_config():
...
+    with open(yatest.common.source_path("ydb/public/tools/lib/cmds/ut/patch.yaml"), 'r') as fh:
+        data_2 = yaml.load(fh, Loader=yaml.FullLoader)
+
+    with open(patched_yaml, "w") as res:
+        res.write(yaml.dump(merge_two_yaml_configs(data_1, data_2), default_flow_style=False))
+
+    return yatest.common.canonical_file(patched_yaml, local=True)
+
+def test_merge_two_dicts():
+    dict_1 = {'data': [{'user': 'root', 'password': '1234'}]}
```
Comment (rekby) [inline ydb/public/tools/lib/cmds/ut/test.py:43]: see to https://docs.pytest.org/en/stable/how-to/parametrize.html#pytest-mark-parametrize-parametrizing-test-functions for test parameters

File: ydb/public/tools/lib/cmds/ut/patch.yaml [hunk 18]
```diff
@@ -0,0 +1,7 @@
+table_service_config:
```
Comment (rekby) [inline ydb/public/tools/lib/cmds/ut/patch.yaml:1]: add some list for demonstrate replacement

**Review by rekby — DISMISSED** [review]

**Review by rekby — APPROVED** [review]
